### PR TITLE
Deploy to GitHub pages

### LINKED ISSUE #39: Publish to gh-pages
Even better if we can make it work with https://github.com/rust-unofficial/too-many-lists/pull/37

### LINKED ISSUE #95: Broken image in 2.4. Push
Indy has gone without anybody noticing that he's gone!
On the website the link to http://cglab.ca/%7Eabeinges/blah/too-many-lists/indy.gif in secion 2.4: Push is broken.

### LINKED ISSUE #103: Pre-rendered version is down
http://cglab.ca/~abeinges/blah/too-many-lists/book/ is giving an error 403 and it's not the first time I've seen it down in the last year or so.

Perhaps a pre-rendered version could be offered via GitHub Pages to improve availability?

### PR DESCRIPTION
**Pull request by lzutao**
Some notable changes in this PR:

* Use gh-pages for deploying this book (Close #103, close #39)
  * A demo is at https://lzutao.github.io/too-many-lists/.
  * Fix path for `indy.gif`: Close #95.
* Use the same mdbook version as rust-lang/book.
  * Migrate to mdBook v0.2.1.
* Change git settings:
   * Change git.depth to 1 (default 50).
   * Make `git clone` quiet.
* Rename build stage:
   * Rename `before_script` stage to `install`.
   * Rename `after_success` stage to `before_deploy`.
* No install Rust toolchain.
   **Rationale**: We don't test any Rust code, so we don't need to install rust
   toolchain. This would reduce waiting time for the real script to
   work.

### CODE REVIEW

**Comment by lzutao** [comment]
This PR ran for about 39 seconds, older build ran for about 9 min 46 sec.

**Comment by lzutao** [comment]
Rebased on master.

**Comment by lzutao** [comment]
ping @edunham, cause Gankro is little bit busy now.

File: .travis.yml [hunk 1]
```diff
@@ -1,27 +1,22 @@
...
-  file: "too-many-lists.zip"
-  skip_cleanup: true
+  provider: pages
+  skip-cleanup: true
+  local-dir: book
+  github-token: $GH_DEPLOY_TOKEN  # Set in the settings page of your repository, as a secure variable
+  keep-history: true
   on:
-    tags: true
+    branch: master
```
Comment (lzutao) [inline .travis.yml]: Should we un-change this? Which means that only deploys the book when having tags.

**Comment by lzutao** [comment]
ping @Manishearth. Could you take a look?

File: .travis.yml (hunk 1 above)
Comment (Manishearth) [reply .travis.yml]: nah master is fine.

**Comment by lzutao** [comment]
@Manishearth Thanks, but it seems like this PR failed to deploy because of wrong invalid gh-token.

**Comment by Manishearth** [comment]
Hmm, there's a token there already, but I'll check

**Comment by lzutao** [comment]
Maybe travis deploy for gh-page needs [Personal access token](https://docs.travis-ci.com/user/deployment/pages/#setting-the-github-token).

File: .travis.yml [hunk 2]
```diff
@@ -1,27 +1,22 @@
...
 
 deploy:
-  provider: releases
-  api_key: "$GH_DEPLOY_TOKEN"
-  file: "too-many-lists.zip"
-  skip_cleanup: true
+  provider: pages
+  skip-cleanup: true
+  local-dir: book
+  github-token: $GH_DEPLOY_TOKEN  # Set in the settings page of your repository, as a secure variable
```
Comment (lzutao) [inline .travis.yml]: This requires [personal access token](https://docs.travis-ci.com/user/deployment/pages/#setting-the-github-token). While it has the same variable name as `api_key`, it is not equivalent. cc @Manishearth

Comment (Manishearth) [reply .travis.yml]: Yeah, I'm aware, the travis thing had one set up, I'll add one.

Comment (Manishearth) [reply .travis.yml]: The old one probably expired.

**Comment by lzutao** [comment]
@Manishearth The build is now completes and deployed to gh-pages branch. But we still need to choose gh-pages branch as deploying method in repo settings.
image

**Comment by Manishearth** [comment]
http://rust-unofficial.github.io/too-many-lists/index.html

It's unclear to me why the readme is still there, probably a caching issue

**Comment by lzutao** [comment]
Good on my side now! Thanks @Manishearth for taking your time to do this.

**Pull request merged**